### PR TITLE
Implement our own wrapper around nats.Conn

### PR DESCRIPTION
**Pull request by DavidS-om**
Since nats.EncodedConn is deprecated and we need more control around header management to inject opentelemetry propagation context.

Core piece is the sdp.EncodedConnection interface and sdp.EncodedConnectionImpl, which take care of propagating the otel trace context and protobuf marshaling.

Everything else is just for connecting all the dots after that change, like threading a ctx through all processing

### CODE REVIEW

File: connection.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+type ProtoRawMsgHandler[M proto.Message] func(ctx context.Context, msg *nats.Msg, m M)
+
+// NewMsgHandler Create a new nats.MsgHandler from a ProtoMsgHandler[M] with a specified proto.Message that takes care of Otel Propagation and Protobuf marshaling
+func NewRawMsgHandler[M proto.Message](spanName string, h ProtoRawMsgHandler[M], alloc func() M, spanOpts ...trace.SpanStartOption) nats.MsgHandler {
+	if h == nil {
+		return nil
+	}
+
+	return NewOtelExtractingHandler(spanName, func(ctx context.Context, msg *nats.Msg) {
+		var payload M = alloc()
```
Comment (DavidS-om) [inline connection.go:132]: I REALLY don't like it, but I couldn't find a better solution for allocating generic memory in a short search.  Any ideas?

Comment (DavidS-om) [reply connection.go:132]: since `proto.Message` is an interface, any hack I could come up with only returns `nil`.

**Review by dylanratcliffe — CHANGES_REQUESTED** [review]
I think we should have `Start()` handle context cancellation too, plus one question about a TODO

File: progress.go [hunk 2]
```diff
@@ -59,17 +50,18 @@ type ResponseSender struct {
...
 // Start sends the first response on the given subject and connection to say
 // that the request is being worked on. It also starts a go routine to continue
 // sending responses until it is cancelled
 //
 // Note that the NATS connection must be an encoded connection that is able to
-// encode and decode SDP messages. This can be done using
-// `nats.RegisterEncoder("sdp", &sdp.ENCODER)`
-func (rs *ResponseSender) Start(natsConnection EncodedConnection, responderName string) {
+// encode and decode SDP messages.
+func (rs *ResponseSender) Start(ctx context.Context, ec EncodedConnection, responderName string) {
```
[the remark this answers is not in the document]
Comment (DavidS-om) [reply progress.go:62]: :heavy_check_mark:

File: progress.go [hunk 3]
```diff
@@ -137,16 +131,17 @@ func (rs *ResponseSender) Kill() {
...
+		// Create the response before starting the goroutine since it only needs to
+		// be done once
+		resp := Response{
+			Responder: rs.responderName,
+			State:     ResponderState_COMPLETE,
+		}
+
+		// Send the final completion message
 		rs.connection.Publish(
+			context.TODO(),
```
[the remark this answers is not in the document]
Comment (DavidS-om) [reply progress.go]: I was hoping to eventually get a ctx passed into `Done()`, but that didn't naturally happen, and I also couldn't find someone calling this in a quick search. Demoted it to Background now.

**Review by dylanratcliffe — APPROVED** [review]
Nice. `KillWithResponse` is a good change too

**Pull request merged**